python/cudf/cudf/core/dataframe.py

```diff
@@ -1337,41 +1337,34 @@ def memory_usage(self, index=True, deep=False):
 
     @annotate("DATAFRAME_ARRAY_FUNCTION", color="blue", domain="cudf_python")
     def __array_function__(self, func, types, args, kwargs):
-
-        cudf_df_module = DataFrame
-        cudf_series_module = Series
-
-        for submodule in func.__module__.split(".")[1:]:
-            # point cudf to the correct submodule
-            if hasattr(cudf_df_module, submodule):
-                cudf_df_module = getattr(cudf_df_module, submodule)
-            else:
-                return NotImplemented
+        if "out" in kwargs or any(
+            not issubclass(t, (Series, DataFrame)) for t in types
+        ):
+            return NotImplemented
 
         fname = func.__name__
 
-        handled_types = [cudf_df_module, cudf_series_module]
-
-        for t in types:
-            if t not in handled_types:
-                return NotImplemented
-
-        if hasattr(cudf_df_module, fname):
-            cudf_func = getattr(cudf_df_module, fname)
-            # Handle case if cudf_func is same as numpy function
-            if cudf_func is func:
-                return NotImplemented
-            # numpy returns an array from the dot product of two dataframes
-            elif (
-                func is np.dot
-                and isinstance(args[0], (DataFrame, pd.DataFrame))
-                and isinstance(args[1], (DataFrame, pd.DataFrame))
-            ):
-                return cudf_func(*args, **kwargs).values
-            else:
-                return cudf_func(*args, **kwargs)
-        else:
-            return NotImplemented
+        try:
+            cudf_func = getattr(self.__class__, fname, None)
+            if cudf_func:
```

Comment:
...wait, no way. A real-world application of the walrus operator? We _do_ require Python 3.8 now. Please accept this refactor. I've waited ages for this moment.

Also remove the variable `fname` above, just inline it here.

```suggestion
            if cudf_func := getattr(self.__class__, func.__name__, None):
```

Reply:
Lol sorry to rain on your parade here. Yes, this is a real-world use case for the operation, and I did consider it, but I decided it would be a good idea not to quietly introduce syntax that was the reason Guido decided to quit Python. IMO any piece of syntax this controversial should not be introduced without more buy-in from the team that it won't adversely affect readability. And tbh I wouldn't be surprised if one of our linters eventually decided to take a stricter stance on the operator. [There's even a flake8 plugin to forbid it!](https://github.com/asottile/flake8-walrus)

Reply:
There are only a small number of cases where I would advocate for using an assignment expression, but I think it's an improvement in readability and brevity here. If you feel strongly about not using assignment expressions, we can leave it. I wouldn't expect any linters to disallow this operator in the future, that seems like a very strong stance to take on an operator.

I would still like to see `fname` removed since the variable is only used once.

Reply:
The scope of the variable persists beyond the `if`. The benefit is to merge two lines, and indicate that the variable being defined in that condition is intended for re-use inside that `if` statement's body. I wouldn't recommend using `:=` in the `if` statement for a variable whose scope is needed beyond the `if` statement body, since that's not very clear.

Reply:
```
python3 -c "
if x := True:
  print(x)

print(x)
"
```
Yeah, so x is still defined and assigned outside of the if statement, so `:=` is gross in this case, imho.

Reply:
I take back what I said about `:=` being gross. In this case it's no worse than just assigning the variable within the if block, since everything is function-scoped. It's the lack of block-scope in general that I find... not great.

Reply:
OK, based on some discussions offline I'm not seeing significant opposition to including this so I'm going to :runner: with the :seal: (:disappointed: there's no walrus emoji, I went with the closest thing).